mentoring/v1/upgrade.py

```diff
@@ -0,0 +1,97 @@
+# -*- coding: utf-8 -*-
+"""
+The mentoring XBlock was previously designed to be edited using XML.
+
+This file contains a script to help migrate mentoring blocks to the new format which is
+optimized for editing in Studio.
+
+To run the script on devstack:
+SERVICE_VARIANT=cms DJANGO_SETTINGS_MODULE="cms.envs.devstack" python -m mentoring.v1.upgrade
+"""
+import logging
+from lxml import etree
+from mentoring import MentoringBlock
+from StringIO import StringIO
+import sys
+from .studio_xml_utils import studio_update_from_node
+from .xml_changes import convert_xml_v1_to_v2
+
+
+def upgrade_block(block):
+    """
+    Given a MentoringBlock "block" with old-style (v1) data in its "xml_content" field, parse
+    the XML and re-create the block with new-style (v2) children and settings.
+    """
+    assert isinstance(block, MentoringBlock)
+    assert bool(block.xml_content)  # If it's a v1 block it will have xml_content
+
+    xml_content_str = block.xml_content
+    parser = etree.XMLParser(remove_blank_text=True)
+    root = etree.parse(StringIO(xml_content_str), parser=parser).getroot()
+    assert root.tag == "mentoring"
+    convert_xml_v1_to_v2(root)
+
+    # We need some special-case handling to deal with HTML being an XModule and not a pure XBlock:
+    try:
+        from xmodule.html_module import HtmlDescriptor
+    except ImportError:
+        pass  # Perhaps HtmlModule has been converted to an XBlock?
+    else:
+        @classmethod
+        def parse_xml_for_HtmlDescriptor(cls, node, runtime, keys, id_generator):
+            block = runtime.construct_xblock_from_class(cls, keys)
+            block.data = node.text if node.text else ""
+            for child in list(node):
+                if isinstance(child.tag, basestring):
+                    block.data += etree.tostring(child)
+            return block
+        HtmlDescriptor.parse_xml = parse_xml_for_HtmlDescriptor
+
+    # Save the xml_content to make this processes rerunnable, in case it doesn't work correctly the first time.
+    root.attrib["xml_content"] = xml_content_str
+
+    # Replace block with the new version and the new children:
+    studio_update_from_node(block, root)
```

Comment:
OK, I've changed the upgrade script to publish the new block if the previous one was published.

Have not yet checked the preservation of student data, though I didn't change the format in which it is stored.